### PR TITLE
Add Encryption Functionality and Tests

### PR DESCRIPTION
**Pull request by ravenstewart**
This PR makes progress towards a solution for issue #10698.

### Description
I added an EncryptionUtils static class which currently has one function:
`public static String getEncryptedAppLog(Context context)`

This function returns a String containing JSON in this format:
```
        {
            "keyedWith": "v1",
            "encryptedKey": "$key_as_base_64",  // The encrypted AES key
            "header": "base_64_encoded_header", // The xchacha20poly1305 stream header
            "messages": []                      // the stream elements, base-64 encoded
        }
```

The `messages` value is a JSON array containing encrypted elements that represent lines in the logs from AppLog.  The last element in the JSON array is an empty string with the `FINAL` tag.

Calling this function will also write out two test output files in the Android device storage:
- `app_log.txt `: the original log text from AppLog
- `app_log_encrypted.json`: JSON with encrypted log data

### Testing Instructions

**Steps:**
1. Call this function from the App after AppLog has accumulated some logs.  (For my tests, the function call was added to the  `void copyAppLogToClipboard()` function in `AppLogViewerActivity.java`.  This way I can view the logs in the app before I generate the test files.)
2. Build the Android app, run the app in the emulator, and trigger the function call to generate the test files.
3. Run the command `adb pull /sdcard/<test file name> <target path>` to copy the files to your hard drive
4. Run the command `curl --data "@<file name>.json" https://log-encryption-testing.herokuapp.com`
5. Compare the output from the test endpoint with the contents of `app_log.txt`

**Expected Results:**
The contents from the test endpoint should be the same as the contents of `app_log.txt`.  (Currently, there is an extra new line at the end of the output from the test endpoint.)

### Next Steps
- Use this function to create encrypted log data associated with Sentry events
    - Need to associate an UUID with the log data and send the UUID with the Sentry Event
    - Where should we store the log data and how will the team access it?
- Use this function to create encrypted log data used for Zendesk ticket creation
    - How will the team decrypt the data once they retrieve it from Zendesk?

### Additional Notes

- I chose the [libsodium-JNI](https://github.com/joshjdevl/libsodium-jni) Java (Android) binding because it had the simplest steps for integration out of all of the bindings for Java (Android) that had the features that we needed.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+            BufferedWriter logOutput = new BufferedWriter(new FileWriter(path + "app_log.txt"));
+            logOutput.write(logText);
+            logOutput.close();
+
+            BufferedWriter encryptedLogOut = new BufferedWriter(new FileWriter(path + "app_log_encrypted.json"));
+            encryptedLogOut.write(encryptedLogJson.toString(4));
+            encryptedLogOut.close();
+
+            ToastUtils.showToast(context, "EncryptionUtils test code, test files saved to: " + path);
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: While I understand that this is still testing code, we can't merge it into the repo as-is, because utility classes shouldn't be able to present UI. I think using unit tests to test this code will resolve this issue :)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        NaCl.sodium().crypto_secretstream_xchacha20poly1305_push(
+            state,
+            encryptedLogLine,
+            clen,
+            emptyString.getBytes(),
+            0,
+            ad,
+            0,
+            (short) NaCl.sodium().crypto_secretstream_xchacha20poly1305_tag_final());
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: It'd be nice to extract this into a class constant for consistency (even though it's only used once)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+            NaCl.sodium().crypto_secretstream_xchacha20poly1305_push(
+                state,
+                encryptedLogLine,
+                clen,
+                logLineBytes,
+                logLineBytes.length,
+                ad,
+                0,
+                (short) 0);
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: It'd be nice to extract this into a class constant for consistency (even though it's only used once)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    /*
+        Returns a JSON String containing following data:
+        {
+            "keyedWith": "v1",
+            "encryptedKey": "$key_as_base_64",  // The encrypted AES key
+            "header": "base_64_encoded_header", // The xchacha20poly1305 stream header
+            "messages": []                      // the stream elements, base-64 encoded
+        }
+    */
+    public static String getEncryptedAppLog(Context context) throws JSONException {
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: This method does a lot of things – could any of it be simplified into smaller methods that can be more easily tested in isolation?

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.libsodium.jni.NaCl;
+
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+
+public class EncryptionUtils {
+    static final int BOX_SEALBYTES = NaCl.sodium().crypto_box_sealbytes();
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: These class constants are a nice tool for keeping things readable!

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    static final int XCHACHA20POLY1305_KEYBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_keybytes();
+    static final int XCHACHA20POLY1305_STATEBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_statebytes();
+    static final int XCHACHA20POLY1305_HEADERBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_headerbytes();
+    static final int XCHACHA20POLY1305_ABYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_abytes();
+    static final int BASE64_ENCODE_FLAGS = Base64.DEFAULT;
+
+    static final String PUBLIC_KEY_BASE64 = "K0y2oQ++gEN00S4CbCH3IYoBIxVF6H86Wz4wi2t2C3M=";
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Can we avoid hard-coding the public key in the class? Different apps implementing this code may wish to use different keys, so it'd be nice to be able to configure this from the calling app without editing this file (assuming it remains in the app code).

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
+package org.wordpress.android.util;
+
+import android.content.Context;
+import android.os.Environment;
+import android.util.Base64;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.libsodium.jni.NaCl;
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: 👍

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 1 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Understood.  I will remove this once I create unit tests for this functionality.

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 2 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Will do! Thank you. :)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 3 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Will do!  :)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 4 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: I will split this up. :)  Thank you!

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 5 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: :D Thank you!

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 6 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Got it!  I will make this change. Thank you! :)

Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: @jkmassel Would [`key_strings.xml`](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/res/values/key_strings.xml) be a good place for us to specify the public key?

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 2 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Made this change here: [`834eb16`](https://github.com/wordpress-mobile/WordPress-Android/pull/10738/commits/834eb1698de3cbfec7bcb6c499a2d0023ba8da72)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 3 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Made this change here: [`834eb16`](https://github.com/wordpress-mobile/WordPress-Android/pull/10738/commits/834eb1698de3cbfec7bcb6c499a2d0023ba8da72)

**Comment by ravenstewart** [comment]
@jkmassel Thank you for taking the time to review this code and write up the helpful feedback!  :)

I will begin making the suggested changes and also look into Tracks to see if it makes sense to move this class over there.

I will also make a note regarding persisting AppLog logs between launches.

**Comment by ravenstewart** [comment]
@jkmassel, here are my thoughts regarding moving the EncryptionUtils class into the Tracks library. :)

I agree that it would be good to move this out of app specific code and allow the encryption functionality to be used by other apps.

I believe the WordPress-Android project is currently not using the Tracks library for logging.  Because of this, it may feel a little bit out of place to access encryption functionality from the Tracks library in order to encrypt logs from AppLog.  But, if the goal is to move towards having the Tracks library handle all logging needs (including the logs needed by Zendesk), then it could make sense for the encryption utility to exist there... If the intention is to eventually have only the Tracks library use the encryption utility and have any app specific code rely on the Tracks library to handle log processing details.

In the future, if the encryption functionality is ever needed for use outside of what Tracks is intended for, we would want to move EncryptionUtils into a more general purpose library or an encryption specific library.

So the short answer is, ideally we may want to decouple encryption functionality from the Tracks library.  But, if only the Tracks library will be using the encryption functionality, having the EncryptionUtils exist there would be ok.

Hope I am understanding the intentions correctly and that this makes sense. :)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 1 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: This has been removed here: 77d4e2eb68ad1c3e87a6c1531c27e41f5d70fe90

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 4 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: @jkmassel I split this up into a few smaller methods here: 77d4e2eb68ad1c3e87a6c1531c27e41f5d70fe90  Please let me know what you think. :)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 6 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: the public key will now be a parameter.  the change is here: 77d4e2eb68ad1c3e87a6c1531c27e41f5d70fe90

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 1 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: this has been removed here: 77d4e2eb68ad1c3e87a6c1531c27e41f5d70fe90

**Comment by ravenstewart** [comment]
@jkmassel Here are some notes regarding my recent commits.

I added a `EncryptionUtilsTest` class here: c7f4539485d6ca9f72c6c4aae7e2ccf98250a836

Inside of the test class, there is a test method that uses the `volley` library to send a POST request to the test endpoint.  This POST request contains data which is encrypted using the `encryptStringData` utility function from `EncryptionUtils`.  The test method then compares the response from the POST request with the unencrypted test data.  The unencrypted test data has a delimiter appended to the end of it in this comparison.  The test fails if the two comparison inputs are not equal. 

Hopefully this was an ok first attempt at it. :)  It's been fun digging into automated testing.

Some thoughts I had regarding this change:
- I wasn't sure whether class constants within the test class was a good way to specify some of the test data.
  - When we are using the `encryptStringData` function outside of the test code, would [`key_strings.xml`](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/res/values/key_strings.xml) be a good spot to specify it?
- Would we want to implement more tests with some edge cases like empty strings or problematic special characters etc.?
- Were there smaller parts of the `encryptStringData` functionality that we would want to test in isolation?  I wasn't sure about the best way to approach this, because I felt that each piece of smaller functionality within this utility method relied heavily on another library.

Also, I may have messed up this pull request. 😅 I was trying to update this branch to make it even with wordpress-mobile:develop... and I ended up accidentally adding a bunch of commits to the pull request. I removed the extra commits from the pull request to at least make it easier to read, but I'm not quite sure whether I did it properly.   

Thank you!  Looking forward to your feedback. 😃

**Comment by jkmassel** [comment]
>In the future, if the encryption functionality is ever needed for use outside of what Tracks is intended for, we would want to move EncryptionUtils into a more general purpose library or an encryption specific library.
>
>So the short answer is, ideally we may want to decouple encryption functionality from the Tracks library. But, if only the Tracks library will be using the encryption functionality, having the EncryptionUtils exist there would be ok.

The idea of having this as its own helper library is interesting! We could certainly do that. To answer your question about where it'd be used, it would be used by Tracks to upload the logs following a crash. Additionally, down the road it's likely that it could be used by any app separately from Tracks. One thing to bear in mind is that Tracks is currently used in every project we work on. WDYT?

File: WordPress/src/androidTest/java/org/wordpress/android/util/EncryptionUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        final Context context = ApplicationProvider.getApplicationContext();
+
+        final String encryptionDataJson = getEncryptionDataJson(END_POINT_TEST_PUBLIC_KEY);
+        if (encryptionDataJson == null) {
+            fail("unable to encrypt test data");
+        }
+
+        final CountDownLatch countDownLatch = new CountDownLatch(1);
+
+        StringRequest postRequest = new StringRequest(
```
Comment (jkmassel) [inline WordPress/src/androidTest/java/org/wordpress/android/util/EncryptionUtilsTest.java]: Can we test this in a way that doesn't require a network request?

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            "keyedWith": "v1",
+            "encryptedKey": "$key_as_base_64",  // The encrypted AES key
+            "header": "base_64_encoded_header", // The xchacha20poly1305 stream header
+            "messages": []                      // the stream elements, base-64 encoded
+        }
+    */
+    public static String encryptStringData(final String publicKeyBase64,
+                                           final String stringData,
+                                           final String delimiter) throws JSONException {
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: While I like how configurable this is, I'm not sure we get a lot of benefit from having a `delimiter` field. IMHO, assuming that we'll break up the file by line is sufficient. If, for some reason, someone decides to encode a file that's extremely long and all on one line, I think it's safe for them to deal with the consequences of that. That should simplify the code in several places :)

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    static final int BOX_SEALBYTES = NaCl.sodium().crypto_box_sealbytes();
+    static final int XCHACHA20POLY1305_KEYBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_keybytes();
+    static final int XCHACHA20POLY1305_STATEBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_statebytes();
+    static final int XCHACHA20POLY1305_HEADERBYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_headerbytes();
+    static final int XCHACHA20POLY1305_ABYTES = NaCl.sodium().crypto_secretstream_xchacha20poly1305_abytes();
+    static final int BASE64_ENCODE_FLAGS = Base64.DEFAULT;
+
+    static final short XCHACHA20POLY1305_TAG_FINAL = 
+            (short) NaCl.sodium().crypto_secretstream_xchacha20poly1305_tag_final();
+    static final short XCHACHA20POLY1305_TAG_0 = (short) 0;
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: This shouldn't be a magic number – it has its own tag name (and should have an associated symbol you can call): `crypto_secretstream_xchacha20poly1305_TAG_MESSAGE`

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+        final String headerBase64 = initSecretStreamXchacha20poly1305(state, key);
+        encryptionDataJson.put("header", headerBase64);
+
+        String[] splitStringData = stringData.split(delimiter);
+        JSONArray encryptedElementsJson = new JSONArray();
+  
+
+        for (String element : splitStringData) {
+            element = element + delimiter; // Add delimiter back to the end of the line
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: This changes the content of the message for the last line, which we probably shouldn't do :)

**Comment by ravenstewart** [comment]
> The idea of having this as its own helper library is interesting! We could certainly do that. To answer your question about where it'd be used, it would be used by Tracks to upload the logs following a crash. Additionally, down the road it's likely that it could be used by any app separately from Tracks. One thing to bear in mind is that Tracks is currently used in every project we work on. WDYT?

@jkmassel Would the Zendesk functionality also be using Tracks to manage the encrypted logs it needs?  If so, I think having the encryption utility live within Tracks would be the way to go until we actually need the encryption utility by an app separately from Tracks down the road. 🙂

File: WordPress/src/androidTest/java/org/wordpress/android/util/EncryptionUtilsTest.java (hunk 8 above)
Comment (ravenstewart) [reply WordPress/src/androidTest/java/org/wordpress/android/util/EncryptionUtilsTest.java]: Yes. 😁  Thank you for clarifying!  I'm kind of relieved that this will be taken out.  😅

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 9 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Got it!  I will remove the delimiter. 🙂

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 10 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Aw man! I totally missed that.  Thank you for pointing that out. 😅

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 11 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: Oh boy.  Thank you for catching that. 😅  Yeesh.  😕

**Comment by ravenstewart** [comment]
> I like the split up methods in `EncryptionUtils` – it makes it a bit easier to see where the pieces are.
I agree!  Thank you for the suggestion. 😁

> The PR itself looks ok – no weird commits in there!
Thank you for checking! phew! 😅

Thank you for looking this over and answering all my questions @jkmassel .  🙂  I will get the requested changes taken care of in the next few days.

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 9 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: removed the delimiter parameter here: 09c15a87bc1be23431b4f78d532b5b72837d780c

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 10 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: now using the tag name with this change: 09c15a87bc1be23431b4f78d532b5b72837d780c

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 11 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: fixed the bug here: 09c15a87bc1be23431b4f78d532b5b72837d780c

File: WordPress/src/androidTest/java/org/wordpress/android/util/EncryptionUtilsTest.java (hunk 8 above)
Comment (ravenstewart) [reply WordPress/src/androidTest/java/org/wordpress/android/util/EncryptionUtilsTest.java]: replaced network request test implementation here: aa875f161ce9cb2c2745b65a0181db1eb516005f

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java (hunk 11 above)
Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: oops. through rewriting my test case I figured out that this actually wasn't a bug. 🙂 unfixed it here: aa875f161ce9cb2c2745b65a0181db1eb516005f

**Comment by ravenstewart** [comment]
@jkmassel I added an empty string test case here: 66e521411bfbdec5e2b4acf4f4a212b4df9cfa5a 

Next, I will be working on more test cases and on moving the test strings to the spot you suggested. 🙂

**Comment by ravenstewart** [comment]
@jkmassel  I added a sample special character test case here: 33fadf14fb20a9f51baf9c4e0c51a7744b1cb036

I decided after looking into it that a sample of special characters would suit us just fine rather than trying to test a larger more complete set.  Please let me know what you think!

I also decided to keep the test strings as class constants for now, since the test cases are still fairly small.  Please let me know if you do not like this idea and I am happy to make the change! 🙂

File: WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        NaCl.sodium().crypto_secretstream_xchacha20poly1305_init_push(state, header, key);
+
+        return Base64.encodeToString(header, BASE64_ENCODE_FLAGS);
+    }
+
+    private static String getSecretStreamXchacha20poly1305EncryptedBase64String(final byte[] state,
+                                                                                final String data,
+                                                                                final short tag) {
+        final int[] clen = new int[0];
+        final byte[] ad = new byte[0];
```
Comment (jkmassel) [inline WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: These variable names are a little unclear – could you give them more descriptive names?

Comment (ravenstewart) [reply WordPress/src/main/java/org/wordpress/android/util/EncryptionUtils.java]: yep!  Thank you for catching that. 🙂  Fixed this in: a57da8fcf15c4634b8d2811c3f79275bc6d2130a

**Comment by ravenstewart** [comment]
> This is really good! One little nitpick, then we're ready to merge!

😮 Whoa!! That's exciting!! 🤩 @jkmassel That last request has been addressed!
